test/napi/lib/workers.js

```diff
@@ -106,3 +123,59 @@ describe("Worker / Root Tagging Tests", () => {
     });
   });
 });
+
+describe("Globals", () => {
+  it("should be able to read an instance global from the main thread", () => {
+    let lookedUpId = addon.get_or_init_thread_id(NaN);
+    assert(!Number.isNaN(lookedUpId));
+    assert.strictEqual(lookedUpId, threadId);
+  });
+
+  it("should be able to store rooted objects in instance globals", () => {
+    addon.stash_global_object();
+    assert.strictEqual(global, addon.unstash_global_object());
+  });
+
+  it("should gracefully panic upon reentrant get_or_try_init", () => {
+    // 1. Global should start out uninitialized
+    assert.strictEqual(null, addon.get_reentrant_value());
+
+    // 2. Re-entrancy should panic
+    try {
+      let result = addon.reentrant_try_init(() => {
+        addon.reentrant_try_init(() => {});
+      });
+      assert.fail("should have panicked on re-entrancy");
+    } catch (expected) {}
```

Comment:
We should also assert that the inner closure was never executed. I haven't tested, but I think it current *is* being executed.

Reply:
Great point. I added the test logic. It's actually not being executed, because the inner `get_or_try_init` calls `tx.cell().pre_init()` during `TryInitTransaction::new()`, which panics before we ever get to calling `run()`.